Add documentation for the log configuration (#45940)
gravitational/teleport: docs/pages/admin-guides/management/admin/logging.mdx

---

> title: Logger configuration

Reviewer: ```suggestion
title: Logger Configuration
```
We use title case in titles (here's the style guide: https://github.com/gravitational/docs/blob/main/docs-contributors/style-guide.md)

> description: Logger configuration for Teleport instance

Reviewer: ```suggestion
description: Explains how to configure the logger on a Teleport instance.
```
Makes the description a sentence that begins with a verb per the style guide

---

> ## Configuration

Reviewer: ```suggestion
```
I don't think there's a need for this H2 here, since there are no other headings, and the title already presents the topic of the guide.

> In the configuration file of the Teleport instance, you can configure the logger's behavior by defining the output

Reviewer: ```suggestion
In the configuration file of a Teleport instance, you can configure the logger's behavior by defining the output
```

> destination, severity level, and output format.
> 
> Sample of `/etc/teleport.yaml` configuration for defining the logger settings.
> 

Reviewer: ```suggestion
destination, severity level, and output format:

```

```yaml
teleport:
  log:
    output: stderr
    severity: INFO
    format:
      output: text
      extra_fields: [caller, level]
```

If the output parameter is not defined or set as empty, `stderr` (aliases `err` or `2`) is used by default.
Other available options for defining the output include `stdout` (aliases `out` or `1`), `syslog` for writing
to the syslog file, or a filepath for direct writing to a log file destination.

> Severity has several levels, which are sorted by increasing priority (means that if `info` is selected, `warn` and `err` also applied):

Reviewer 2: ```suggestion
Severity has several levels, which are sorted by decreasing priority (means that if `info` is selected, `warn` and `err` also applied):
```

Author: slog misleads, they have different ordering with logrus
```
slog.LevelError = 8
slog.LevelWarn = 4

logrus.ErrorLevel = 2
logrus.WarnLevel = 3
```

> Severity has several levels, which are sorted by decreasing priority (means that if `info` is selected, `warn` and `err` also applied):

Reviewer: ```
(means that if `info` is selected, `warn` and `err` also applied)
```

We should fix the grammar here and make this a complete sentence. I think it would make sense to separate this into its own sentence rather than use parentheses.

>  - `err`, `error` - used for errors that should definitely be noted.

Reviewer: What does "should definitely be noted" mean?

Author: Means that these log messages indicate actions are required to fix them, errors may lead to an unpredictable state of the application

Reviewer: I would say, "Logs that require action from the user" instead of "Should definitely be noted".

 - `warn`, `warning` - non-critical entries that deserve attention.
 - `info` or empty value - general operational entries about what's going on inside the application.
 - `debug` - usually only enabled when debugging, verbose logging.

>  - `trace` - designates more detailed informational events than the debug.

Reviewer: The descriptions here are self evident, I think. Is there anything we can say about the kinds of logs a user can expect when enabling `debug` or `trace`-level logging?

Author: `trace` usually shows very detail information like sequence of method executions 
Did simple search by code base, so we don't use this level that much:
```golang
s.log.WithField("event", event).Trace("sessionCache: Received watcher event")
s.log.WithField("session_id", s.ID()).Trace("session will be recorded at proxy")
log.WithFields(log.Fields{
			"src_addr":     fmt.Sprintf("%v", source),
			"dst_addr":     fmt.Sprintf("%v", destination),
			"cluster_name": p.clusterName}).Trace("Successfully generated signed PROXY header")
// Okta plugin.
log.Trace("Collating plugin resource")
log.Trace("Installing plugin")
```

> The default format for log output is `text`. Another available format is `json`, which may simplify log

Reviewer: What does "may simplify" mean here?

Author: means that json already has a structure for parsing with field names, if we use text, usually regular expression is used to extract data from text message

parsing for systems like Logstash, Loki, or other log aggregators.

Format `extra_fields` defines additional fields which must be added to the log output:
 - `level` is the log field that stores the verbosity.
 - `component` is the log field that stores the calling component.
 - `caller` is the log field that stores the calling file and line number.
 - `timestamp` is the field that stores the timestamp the log was emitted.

On systemd-based distributions you can watch the log output by running the following command:

```code
$ teleport install systemd -o /etc/systemd/system/teleport.service
$ systemctl enable teleport
$ journalctl -fu teleport
```

To store logs as a file, the filepath should be set in the `log.output` configuration.

```yaml
teleport:
  log:
    output: /var/lib/teleport/log/output.log
```

> 

Reviewer 3: I would organize this into a separate sub-section by giving it a title.
```suggestion

## Log rotation support

```

> When Teleport open or create a new log file, a filesystem watcher is launched in the background to monitor modifications.

Reviewer 3: ```suggestion
When Teleport opens or creates a new log file, a filesystem watcher is launched in the background to monitor the file modifications.
```

> If the file is renamed, moved, or deleted, a new one is created using the path defined in the configuration.

Reviewer 3: ```suggestion
If the log file is renamed, moved, or deleted, Teleport automatically creates a new one.
```

This is useful for implementing log rotation without needing to restart or interrupt the main service.

> In this case, the log rotation configuration might look like the following sample `/etc/logrotate.d/teleport.conf`:

Reviewer 3: ```suggestion
Using `logrotate` as an example, you may define the following config `/etc/logrotate.d/teleport.conf` to rotate Teleport log file weekly:
```
